<a href="https://colab.research.google.com/github/abnd-me/Public/blob/main/Abhinand_p_p.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###Read Dataset

In [2]:
import pandas as pd

df = pd.read_csv('/content/titanic_dataset .csv')

df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [3]:
df.set_index('PassengerId' , inplace=True) #make passengerID as index column

In [4]:
df.isna().sum()  #to find null value

Survived      0
Pclass        0
Name          0
Sex           0
Age         177
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       687
Embarked      2
dtype: int64

In [5]:
df.isna().sum()

Survived      0
Pclass        0
Name          0
Sex           0
Age         177
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       687
Embarked      2
dtype: int64

#Preprocessing

###Handel Missing value

In [ ]:
#Handel mising values for numerical columns

df1 = df

df2 = df1.fillna(df1.mean())

In [7]:
df2.isna().sum()

Survived      0
Pclass        0
Name          0
Sex           0
Age           0
SibSp         0
Parch         0
Ticket        0
Fare          0
Cabin       687
Embarked      2
dtype: int64

In [8]:
#Handle missing values for categorical columns

df2['Embarked'].fillna(df2['Embarked'].mode()[0] , inplace = True)  #using mode

constant_value = 0
df2['Cabin'].fillna(constant_value, inplace=True )   #using constant value

df2.isna().sum()

Survived    0
Pclass      0
Name        0
Sex         0
Age         0
SibSp       0
Parch       0
Ticket      0
Fare        0
Cabin       0
Embarked    0
dtype: int64

###Scale the dataset

In [9]:
#Scale Numrical Columns using MinMaxScaler

from sklearn.preprocessing import MinMaxScaler

columns_to_scale = ['Age', 'SibSp', 'Parch', 'Fare']

# Applying MinMax scaling
scaler = MinMaxScaler()
df2[columns_to_scale] = scaler.fit_transform(df2[columns_to_scale])

# Displaying the scaled DataFrame
df2.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex  \
PassengerId                                                              
1                                      Braund, Mr. Owen Harris    male   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
3                                       Heikkinen, Miss. Laina  female   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
5                                     Allen, Mr. William Henry    male   

                  Age  SibSp  Parch            Ticket      Fare Cabin Embarked  
PassengerId                                                                     
1            0.271174  0.125    0.0         A/5 21171  0.014151     0        S  
2            0.472229  0.125    0.0          PC 17599  0.139136   C85        C  
3            0.321438  0.000    0.0  STON/O2. 3101282  0.015469     0        S  
4            0.434531  0.125    0.0            113803  0.103644  C123        S  
5            0.434531  0.000    0.0            373450  0.015713     0        S

## KNN And SVM Model

In [10]:
#train test split
from sklearn.model_selection import train_test_split


features = ['Age', 'SibSp', 'Parch', 'Fare']


X = df2[features] #feature data

y = df2['Survived'] #target data

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [11]:
#Apply KNN and SVM model
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC

#Creting Model
knn_model = KNeighborsClassifier()
svm_model = SVC()

In [12]:
#train svm model
svm_model.fit(X_train , y_train)

SVC()

In [13]:
#train knn model
knn_model.fit(X_train , y_train)

KNeighborsClassifier()

In [14]:
#prediction

knn_pred = knn_model.predict(X_test)
svm_pred = svm_model.predict(X_test)

In [15]:
#Check Accuracy of both model

from sklearn.metrics import accuracy_score

knn_acc = accuracy_score(y_test, knn_pred)
svm_acc = accuracy_score(y_test, svm_pred)

print('Accuracy Of KNN is', knn_acc)
print('Accuracy Of SVM is', svm_acc)

Accuracy Of KNN is 0.6927374301675978
Accuracy Of SVM is 0.7150837988826816


#KFold Cross Validation

In [16]:
from sklearn.model_selection import KFold, cross_val_score

#KFold validation on KNN
kfold_val = KFold(n_splits=10)
kfold_score_knn = cross_val_score(knn_model, X, y, cv=kfold_val)

kfold_score_knn

array([0.65555556, 0.70786517, 0.6741573 , 0.66292135, 0.65168539,
       0.70786517, 0.61797753, 0.71910112, 0.73033708, 0.83146067])

In [18]:
#Average accuracy

avg_kfold_score_knn = kfold_score_knn.mean()

print('KFold Validation accuracy for KNN model is',avg_kfold_score_knn)

KFold Validation accuracy for KNN model is 0.6958926342072409


In [19]:
#KFold validation on SVM

kfold_score_svm = cross_val_score(svm_model, X, y, cv=kfold_val)

kfold_score_svm

array([0.62222222, 0.68539326, 0.74157303, 0.70786517, 0.68539326,
       0.78651685, 0.71910112, 0.74157303, 0.78651685, 0.75280899])

In [20]:
#Average accuracy

avg_kfold_score_svm = kfold_score_svm.mean()

print('KFold Validation accuracy for SVM model is',avg_kfold_score_svm)

KFold Validation accuracy for SVM model is 0.7228963795255929


#Stratified KFold Cross Validation

In [21]:
from sklearn.model_selection import StratifiedKFold

#Apply stratified KFold
stratified_kfold = StratifiedKFold(n_splits=6,)

#accuracy of both validation
str_kfold_score_knn = cross_val_score(knn_model, X, y, cv=stratified_kfold)

str_kfold_score_svm = cross_val_score(svm_model, X, y, cv=stratified_kfold)

In [22]:
# Stratified KFold accuracy score

#KNN
str_kfold_score_knn


array([0.6442953 , 0.69798658, 0.66442953, 0.66891892, 0.7027027 ,
       0.76351351])

In [23]:
# Stratified KFold accuracy score

#SVM
str_kfold_score_svm

array([0.61744966, 0.71812081, 0.73825503, 0.72972973, 0.74324324,
       0.76351351])

In [24]:
# Stratified KFold mean accuracy score


avg_str_kfold_score_knn = str_kfold_score_knn.mean()
avg_str_kfold_score_svm = str_kfold_score_svm.mean()


print('Stratified KFold Validation accuracy for KNN model is',avg_str_kfold_score_knn)
print('Stratified KFold Validation accuracy for SVM model is',avg_str_kfold_score_svm)




Stratified KFold Validation accuracy for KNN model is 0.6903077574218514
Stratified KFold Validation accuracy for SVM model is 0.7183853316403651


 ## When comparing both cross validation we can see that KFold gives accuracy of 69.58%,72.28% for KNN and SVM models respectively.
 ## And Stratified KFold gives an  accuracy of 69.03%,71.83% for KNN and SVM models respecitively.
 ## So here we can say that KFold with 10 splits give slightly better accuracy than Stratified KFold with 6 splits.
